### PR TITLE
Enable daily metrics models backfill of historical data in platform, but not progress

### PR DESCRIPTION
**Pull request by johnbaldwin**
https://appsembler.atlassian.net/browse/BLACK-2375

Porting ad-hoc scripts into Figures to fill approximate metrics for historical values

There should be one more commit, modify the existing Django management command, `backfill_figures_daily_metrics` and add test coverage for that command

## Commit 8aeced479a0949eef179c24ed17b5db2707e8d46

Add method - figures.course.Course.first_enrollment_timestamp

This exists to help identify when we should include a course in
performing backfill to `CourseDailyMetrics` for a given date in the
past.

Getting the `CourseEnrollment.created` value for the first enrollment
provides a fast "close enough best guess" when a course was first
created, as we cannot rely on CourseOverview timestamps and this avoids
diving into MongoDB or the filesystem to dredge up timestamps for the
given course

## Commit  7146aa49153ee7ecf69be1eba09ea6246ba296f1

Commit message has the important details, but here's a brief summary:

This one fixes a latent bug in an existing test case that appeared because the test case did not explicitly set course enrollment 'created' dates for the `CourseEnrollmentFactory` instantiations. After we added new CourseEnrollmentFactory created instances, the `created` date got bumped past the date check. 

## Commit 8aeced479a0949eef179c24ed17b5db2707e8d46

New functions to call backfill for a site and date and track the backfill to a log file

## Commit 0fa88808f6e166f10016a61959aaf66270898cc4

Add new test module, remove test class from old test module

### CODE REVIEW

**Comment by johnbaldwin** [comment]
Interesting. Adding the code causes the following to fail

```
tests/pipeline/test_course_daily_metrics.py ......F............s

 =================================== FAILURES ===================================
___ TestCourseDailyMetricsPipelineFunctions.test_get_num_learners_completed ____

self = <tests.pipeline.test_course_daily_metrics.TestCourseDailyMetricsPipelineFunctions object at 0x7f443a6d7e90>

    def test_get_num_learners_completed(self):
        actual = pipeline_cdm.get_num_learners_completed(
            course_id=self.course_overview.id,
            date_for=self.today)
>       assert actual == len(self.generated_certificates)
E       assert 2 == 3
```

**Comment by johnbaldwin** [comment]
@OmarIthawi @melvinsoft @shadinaif @estherjsuh @bryanlandia This is ready for review

**Comment by johnbaldwin** [comment]
@OmarIthawi @melvinsoft @shadinaif @estherjsuh @bryanlandia Following up, would someone please review?

File: tests/pipeline/test_backfill_daily_metrics.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        monkeypatch.setattr('figures.pipeline.backfill.get_course_keys_for_site',
+                            lambda site: [obj.id for obj in course_overviews])
+        monkeypatch.setattr('figures.pipeline.course_daily_metrics.figures.sites.get_site_for_course',
+                            lambda course_id: self.site)
+        with mock.patch('figures.pipeline.backfill.open', open_mock):
+            results = backfill_daily_metrics_for_site_and_date(
+                site=self.site,
+                date_for=date_for)
+
+        open_mock.assert_called_with(results['logfile'], 'a', encoding='utf-8')
```
Comment (bryanlandia) [inline tests/pipeline/test_backfill_daily_metrics.py:84]: if you did change from append to overwrite then update here too

**Review by bryanlandia — APPROVED** [review]
I left a few comments but approving... .not the most thorough review but it all looks like it makes sense and has good test coverage 🎉

File: figures/pipeline/backfill.py [hunk 2]
```diff
@@ -4,19 +4,53 @@
...
+    Initially created to raise an exception if cached data or data provided as
+    function parameters do not belong to the site being processed.
+    """
+    pass
+
+
+def figures_backfill_log_dir():
+    """Specify where to write Figures backfill log files
+    """
+    return settings.ENV_TOKENS['FIGURES'].get('BACKFILL_LOG_DIR',
```
[the remark this answers is not in the document]
Comment (johnbaldwin) [reply figures/pipeline/backfill.py:52]: @bryanlandia `settings.ENV_TOKENS['FIGURES']` should always exist when Figures is instantiated as an Open edX plugin. Please see here: https://github.com/appsembler/figures/blob/main/figures/settings/lms_production.py#L127

What I did was create an issue to note that you raised the issue and as a place to discuss how to improve this architecture: https://github.com/appsembler/figures/issues/456

File: figures/pipeline/backfill.py [hunk 3]
```diff
@@ -86,3 +120,136 @@ def backfill_enrollment_data_for_site(site):
...
+        logdir = figures_backfill_log_dir()
+
+    filename = 'backfill-for-site-{site_id}-date-{date_for}.log'.format(
+        site_id=site.id, date_for=date_for_str)
+    filepath = os.path.join(logdir, filename)
+
+    course_ids = courses_enrolled_on_or_before(site, date_for)
+    course_id_count = len(course_ids)
+
+    with open(filepath, 'a', encoding='utf-8') as logfile:
```
[the remark this answers is not in the document]
Comment (johnbaldwin) [reply figures/pipeline/backfill.py:216]: This is a log file that has the site-id and date-for as the unique identifiers. Setting to append is a safe and simple way to make sure that we don't wipe and overwrite an existing log in the event an exception gets raised and we have to manually fix something and restart the backfill for that site and date.

A obvious improvement is to rename an exiting log to say dot some number or dot timestamp. But this is an improvement that is not essential at this time. If we find it a pain point we can improve it

File: tests/pipeline/test_backfill_daily_metrics.py (hunk 1 above)
Comment (johnbaldwin) [reply tests/pipeline/test_backfill_daily_metrics.py:84]: Good point if we did. I'll make inline comments that address your comments

**Comment by johnbaldwin** [comment]
Thanks for reviewing @bryanlandia!

**Pull request merged**